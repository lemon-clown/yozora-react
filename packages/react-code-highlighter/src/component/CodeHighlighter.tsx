import PropTypes from 'prop-types'
import React from 'react'
import vscDarkTheme from '../theme/vsc-dark'
import vscLightTheme from '../theme/vsc-light'
import type { IPrismTheme } from '../types/prism'
import { HighlightContent } from './HighlightContent'

export interface ICodeHighlighterProps {
  /**
   * Source code contents
   */
  value: string
  /**
   * Code language
   */
  lang?: string | null
  /**
   * Line number of Lines that should be highlighted.
   */
  highlightLinenos?: number[]
  /**
   * Whether the code block is in a collapsed state.
   * @default false
   */
  collapsed?: boolean
  /**
   * Maximum number of rows displayed
   * @default -1
   */
  maxLines?: number
  /**
   * Line height.
   * @default '1.6rem'
   */
  lineHeight?: React.CSSProperties['lineHeight']
  /**
   * Whether should display line numbers.
   * @default true
   */
  showLineNo?: boolean
  /**
   * If true, use vscDarkTheme as default theme,
   * otherwise use vscLightTheme as default theme.
   * @default true
   */
  darken?: boolean
  /**
   * Highlight prism theme.
   */
  theme?: IPrismTheme
  /**
   * Ref of the codes area.
   */
  codesRef?: React.RefCallback<HTMLDivElement> | React.RefObject<HTMLDivElement>
  /**
   * Callback when linenoWidth changed.
   */
  onLinenoWidthChange?(linenoWidth: React.CSSProperties['width']): void
}

export const CodeHighlighter: React.FC<ICodeHighlighterProps> = props => {
  const {
    lang,
    value: code,
    darken = true,
    highlightLinenos,
    maxLines = -1,
    lineHeight = '1.6rem',
    collapsed = false,
    showLineNo = true,
    codesRef,
    onLinenoWidthChange,
  } = props

  const theme: IPrismTheme = props.theme ?? (darken ? vscDarkTheme : vscLightTheme)
  return (
    <HighlightContent
      code={code}
      codesRef={codesRef}
      collapsed={collapsed}
      highlightLinenos={highlightLinenos}
      lang={lang ?? undefined}
      lineHeight={lineHeight}
      maxLines={maxLines}
      showLineNo={showLineNo}
      theme={theme}
      onLinenoWidthChange={onLinenoWidthChange}
    />
  )
}

CodeHighlighter.propTypes = {
  codesRef: PropTypes.any,
  collapsed: PropTypes.bool,
  darken: PropTypes.bool,
  highlightLinenos: PropTypes.arrayOf(PropTypes.number) as React.Validator<
    number[] | null | undefined
  >,
  lang: PropTypes.string,
  lineHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  maxLines: PropTypes.number,
  onLinenoWidthChange: PropTypes.func,
  showLineNo: PropTypes.bool,
  theme: PropTypes.any,
  value: PropTypes.string.isRequired,
}
CodeHighlighter.displayName = 'YozoraCodeHighlighter'
