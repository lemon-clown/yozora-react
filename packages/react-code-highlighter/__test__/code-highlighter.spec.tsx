import { render } from '@testing-library/react'
import React, { useEffect, useState } from 'react'
import CodeHighlighter from '../src'

describe('basic rendering case', () => {
  test('lineno change', () => {
    function Wrapper(): React.ReactElement {
      const [code, setCode] = useState<string>('let a: number = 1 + 2;')
      const countOfLine: number = code.split(/\r\n|\n|\r/g).length

      useEffect(() => {
        const nextCode =
          'let a = 1, b = 2\n' +
          Array.from(new Array(100))
            .map((_x, i) => '// ' + i)
            .join('\n') +
          "\nlet c = 3\nconsole.log('c:', c)"
        setCode(nextCode)
      }, [])

      return (
        <pre data-testid="pre" data-line-count={countOfLine}>
          <CodeHighlighter lang="typescript" value={code} />
        </pre>
      )
    }

    const view = render(<Wrapper />)
    expect(view.getByTestId('pre')).toHaveAttribute('data-line-count', String(103))
  })

  test('snapshot', () => {
    const view = render(
      <pre>
        <CodeHighlighter lang="typescript" value="let a: number = 1 + 2;" />
      </pre>,
    )
    expect(view.asFragment()).toMatchSnapshot()
  })
})
