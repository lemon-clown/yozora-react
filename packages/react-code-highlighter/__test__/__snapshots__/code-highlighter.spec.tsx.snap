// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic rendering case snapshot 1`] = `
<DocumentFragment>
  <pre>
    <div
      class="css-1k8659m prism-code language-typescript"
      style="color: rgb(212, 212, 212); line-height: 1.6rem; max-height: 3.2rem; min-height: 100%;"
    >
      <div
        class="css-1bwxi5h"
        style="width: 2.2em;"
      >
        <div
          class="css-17fnzo2"
        >
          <span>
            1
          </span>
        </div>
      </div>
      <div
        class="css-rv5jjl"
      >
        <div
          class="css-1ttfemb"
        >
          <div
            class="token-line css-7zjwmr"
            style="color: rgb(212, 212, 212);"
          >
            <span
              class="token keyword"
              style="color: rgb(86, 156, 214);"
            >
              let
            </span>
            <span
              class="token plain"
            >
               a
            </span>
            <span
              class="token operator"
              style="color: rgb(212, 212, 212);"
            >
              :
            </span>
            <span
              class="token plain"
            >
               
            </span>
            <span
              class="token builtin"
              style="color: rgb(79, 193, 255);"
            >
              number
            </span>
            <span
              class="token plain"
            >
               
            </span>
            <span
              class="token operator"
              style="color: rgb(212, 212, 212);"
            >
              =
            </span>
            <span
              class="token plain"
            >
               
            </span>
            <span
              class="token number"
              style="color: rgb(181, 206, 168);"
            >
              1
            </span>
            <span
              class="token plain"
            >
               
            </span>
            <span
              class="token operator"
              style="color: rgb(212, 212, 212);"
            >
              +
            </span>
            <span
              class="token plain"
            >
               
            </span>
            <span
              class="token number"
              style="color: rgb(181, 206, 168);"
            >
              2
            </span>
            <span
              class="token punctuation"
              style="color: rgb(106, 153, 85);"
            >
              ;
            </span>
          </div>
        </div>
      </div>
    </div>
  </pre>
</DocumentFragment>
`;
